core/state/trie_prefetcher.go

```diff
@@ -77,15 +112,91 @@ func newTriePrefetcher(db Database, root common.Hash, namespace string) *triePre
 		storageWasteMeter: metrics.GetOrRegisterMeter(prefix+"/storage/waste", nil),
 	}
 	go p.abortLoop()
+	go p.mainLoop()
 	return p
 }
 
+func (p *triePrefetcher) mainLoop() {
+	for {
+		select {
+		case <-p.copyChan:
+			fetcherCopied := &triePrefetcher{
+				db:      p.db,
+				root:    p.root,
+				fetches: make(map[common.Hash]Trie, len(p.fetchers)),
+			}
+			// we're copying an active fetcher, retrieve the current states
+			for root, fetcher := range p.fetchers {
+				fetcherCopied.fetches[root] = fetcher.peek()
+			}
+			p.copyDoneChan <- fetcherCopied
+
+		case pMsg := <-p.prefetchChan:
+			fetcher := p.fetchers[pMsg.root]
+			if fetcher == nil {
+				fetcher = newSubfetcher(p.db, pMsg.root, pMsg.accountHash)
+				p.fetchers[pMsg.root] = fetcher
+			}
+			fetcher.schedule(pMsg.keys)
+
+		case tireMsg := <-p.trieChan:
+			tireMsg.resultChan <- p.fetchers[tireMsg.root]
+
+		case uMsg := <-p.usedChan:
+			if fetcher := p.fetchers[uMsg.root]; fetcher != nil {
+				fetcher.used = uMsg.used
+			}
+
+		case <-p.closeMainChan:
+			for _, fetcher := range p.fetchers {
+				p.abortChan <- fetcher // safe to do multiple times
+				<-fetcher.term
+				if metrics.EnabledExpensive {
+					if fetcher.root == p.root {
+						p.accountLoadMeter.Mark(int64(len(fetcher.seen)))
+						p.accountDupMeter.Mark(int64(fetcher.dups))
+						p.accountSkipMeter.Mark(int64(len(fetcher.tasks)))
+
+						for _, key := range fetcher.used {
+							delete(fetcher.seen, string(key))
+						}
+						p.accountWasteMeter.Mark(int64(len(fetcher.seen)))
+					} else {
+						p.storageLoadMeter.Mark(int64(len(fetcher.seen)))
+						p.storageDupMeter.Mark(int64(fetcher.dups))
+						p.storageSkipMeter.Mark(int64(len(fetcher.tasks)))
+
+						for _, key := range fetcher.used {
+							delete(fetcher.seen, string(key))
+						}
+						p.storageWasteMeter.Mark(int64(len(fetcher.seen)))
+					}
+				}
+			}
+			close(p.closeAbortChan)
+			close(p.closeMainDoneChan)
+			p.fetchers = nil
+			// drain all the channels before quit the loop
+			for {
+				select {
+				case <-p.copyChan:
+				case <-p.prefetchChan:
+				case <-p.trieChan:
+				case <-p.usedChan:
+				default:
+					return
+				}
+			}
+		}
+	}
+}
+
 func (p *triePrefetcher) abortLoop() {
 	for {
 		select {
 		case fetcher := <-p.abortChan:
 			fetcher.abort()
-		case <-p.closeChan:
+		case <-p.closeAbortChan:
 			// drain fetcher channel
```

Comment:
why drain the channel ? before `closeAbortChan` has check `term` for all fetchers.

Reply:
accept, can be removed.

Reply:
accept, can be removed.